### PR TITLE
[refactor][Lang] Unify SNode and Tensor meta data accessor in both Taichi&Python-scope

### PR DESCRIPTION
**Pull request by archibate**
Related issue = #... (if any)

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

### CODE REVIEW

File: docs/scalar_tensor.rst [hunk 1]
```diff
@@ -72,7 +72,82 @@ Declaration
...
+.. function:: a.parent()
+
+    :parameter a: (Tensor) the tensor
+    :return: (SNode) the SNode where ``a`` is placed
+
+    ::
+        x = ti.var(ti.i32)
+        y = ti.var(ti.i32)
+        blk1 = ti.root.dense(ti.ij, (6, 5))
+        blk2 = blk2.dense(ti.ij, (3, 2))
```
Comment (KLozes) [inline docs/scalar_tensor.rst:146]: blk2 = blk1.dense(ti.ij, (3, 2))

File: docs/scalar_tensor.rst [hunk 2]
```diff
@@ -72,7 +72,82 @@ Declaration
...
+        x.shape()  # (6, 5)
+
+        y = ti.var(ti.i32, 6)
+        y.shape()  # (6,)
+
+        z = ti.var(ti.i32, ())
+        z.shape()  # ()
+
+
+.. function:: a.parent()
```
Comment (KLozes) [inline docs/scalar_tensor.rst:137]: Lets also show grandparent syntax. Or are we talking about python scope parent()? Whivh does not have grantparent access yet. We should probably make it consistent with taichi scope parent() . 

```
.. function:: a.parent(n)

    :parameter a: (Tensor) the tensor
    :parameter n: (scalar) the number of parent steps, ie. parent, grandparent, etc...
    :return: (SNode)  the parent SNode of ``a``

    ::
        x = ti.var(ti.i32)
        y = ti.var(ti.i32)
        blk1 = ti.root.dense(ti.ij, (6, 5))
        blk2 = blk1.dense(ti.ij, (3, 2))
        blk1.place(x)
        blk2.place(y)

        x.parent()  # blk1
        y.parent()  # blk2
        y.parent(2) # blk1
```

File: docs/scalar_tensor.rst [hunk 3]
```diff
@@ -72,7 +72,82 @@ Declaration
...
+        x.dim()  # 2
+
+        y = ti.var(ti.i32, 6)
+        y.dim()  # 1
+
+        z = ti.var(ti.i32, ())
+        z.dim()  # 0
+
+
+.. function:: a.shape()
```
Comment (KLozes) [inline docs/scalar_tensor.rst:120]: This one is actually called get_shape() in python scope.

File: docs/scalar_tensor.rst [hunk 4]
```diff
@@ -72,7 +72,82 @@ Declaration
         # ERROR: cannot create tensor after any tensor accesses from the Python-scope!
 
 
-Attribute
+Accessing components
+--------------------
+
+.. attribute:: a[p, q, ...]
+
+    :parameter a: (Tensor) the tensor of scalar
```
Comment (KLozes) [inline docs/scalar_tensor.rst:80]: lets use "scalars" instead of "scalar"

File: docs/scalar_tensor.rst (hunk 3 above)
Comment (archibate) [reply docs/scalar_tensor.rst:120]: `get_shape()` is a method of `SNode`, in `Expr` it's `shape()`...please check L299 of `expr.py` to verify. We may want to refactor to make `SNode` and `Expr` have the same API later in a separate PR.

File: docs/scalar_tensor.rst [hunk 5]
```diff
@@ -72,7 +72,84 @@ Declaration
...
+        y.shape()  # (6,)
+
+        z = ti.var(ti.i32, ())
+        z.shape()  # ()
+
+
+.. function:: a.parent(n = 1)
+
+    :parameter a: (Tensor) the tensor
+    :parameter n: (optional, scalar) the number of parent steps, i.e. parent, grandparent, etc...
```
Comment (yuanming-hu) [inline docs/scalar_tensor.rst:151]:
```suggestion
    :parameter n: (optional, scalar) the number of parent steps. I.e., ``n=1`` for parent, ``n=2`` for grandparent, etc...
```

File: docs/scalar_tensor.rst [hunk 6]
```diff
@@ -72,7 +72,84 @@ Declaration
...
+    :return: (SNode) the parent SNode where ``a`` is placed
```
Comment (yuanming-hu) [inline docs/scalar_tensor.rst:152]:
```suggestion
    :return: (SNode) the parent of ``a``'s containing SNode
```

File: python/taichi/lang/snode.py [hunk 7]
```diff
@@ -57,6 +57,13 @@ def data_type(self):
     def dim(self):
         return self.ptr.num_active_indices()
 
+    def shape(self):
+        dim = self.dim()
```
Comment (k-ye) [inline python/taichi/lang/snode.py:61]:
```py
return tuple([self.get_shape(i) for i in range(self.dim())])
```

File: docs/scalar_tensor.rst [hunk 8]
```diff
@@ -72,7 +72,95 @@ Declaration
         # ERROR: cannot create tensor after any tensor accesses from the Python-scope!
 
 
-Attribute
+Accessing components
+--------------------
+
+.. attribute:: a[p, q, ...]
```
Comment (k-ye) [inline docs/scalar_tensor.rst:78]: nit: When written like this, it gives the reader an impression that a tensor must have at least two dimensions. I feel like this is such a common operation that we don't need to document it very formally. Maybe we should just show some examples? FYI here's `numpy`'s array indexing doc: https://numpy.org/devdocs/reference/arrays.indexing.html#indexing. It's too fancy for us, but it only shows examples...

Comment (archibate) [reply docs/scalar_tensor.rst:78]: What should we do then? To write `a[obj]` and say `obj` is a `tuple`?

Comment (k-ye) [reply docs/scalar_tensor.rst:78]: Maybe just say "You can access an element of the Taichi tensor by an index", then shows your examples?

Comment (archibate) [reply docs/scalar_tensor.rst:78]: Ack for adding a brief desc. But note that this is put in `API Reference` section, if user want a more tutorial version, they will go for `tensor_matrix.rst` in `Overview` section.

File: docs/scalar_tensor.rst [hunk 9]
```diff
@@ -72,7 +72,95 @@ Declaration
...
+
+    :parameter a: (Tensor) the tensor of scalars
+    :parameter p: (scalar) index of the first tensor dimension
+    :parameter q: (scalar) index of the second tensor dimension
+    :return: (scalar) the element at ``[p, q, ...]``
+
+    This extracts the element value at ``[3, 4]`` of tensor ``a``:
+    ::
+
+        x = a[3, 4]
```
Comment (archibate) [inline docs/scalar_tensor.rst:88]:
```suggestion
    This extracts the element value at index ``[3, 4]`` of tensor ``a``:
    ::

        x = a[3, 4]
```

File: docs/scalar_tensor.rst [hunk 10]
```diff
@@ -72,7 +72,95 @@ Declaration
...
+
+    This sets the element value at ``[3, 4]`` of tensor ``a`` to ``2``:
+    ::
+
+        a[3, 4] = 2
+
```
Comment (archibate) [inline docs/scalar_tensor.rst:94]:
```suggestion

    .. note ::

        In Python, x[(exp1, exp2, ..., expN)] is equivalent to x[exp1, exp2, ..., expN]; the latter is just syntactic sugar for the former.

```

File: docs/scalar_tensor.rst (hunk 8 above)
Comment (k-ye) [reply docs/scalar_tensor.rst:80]: > To write a[obj] and say obj is a tuple

Then let's do this, since it's more generic.

**Comment by archibate** [comment]
> Then let's do this, since it's more generic.

We just say `obj is scalar or tuple`?? Many docs like `matrix.rst` are using `[p, q, ..]` style, not sure if you want this in a separate PR.

**Comment by archibate** [comment]
> I agree this part needs some refactoring in a later PR...

I did that in this PR, and now the docs became the 附带品 come with it, is that ok?

> Then let's do this, since it's more generic.

Feel free to do the doc refactor for `x[obj]` in a separate PR. I'll merge this once @yuanming-hu approved my `Unify Taichi&Python-scope`.

**Comment by k-ye** [comment]
> Many docs like matrix.rst are using [p, q, ..] style, not sure if you want this in a separate PR.

I don't remember being looped into the `matrix.rst` doc, and i wasn't aware of that. But it sounds good to me to clean all these up in another PR.

Just to explain my thoughts: when we write `x[p, q, ...]`, it sends a clear signal to the users that `x` and `y` are the required parameters. (Think about how C++11's variadic args template matches the function based on the number of args) Then `x[None]` could become confusing to the users. Of course, since `[]` is such a common operation, I don't expect people to be actually confused.

At last, one difference between matrix and tensor is that, `Matrix` is indeed 2D, so `[p, q]` actually makes sense for this type.

**Pull request merged**